### ISSUE TITLE
Feature Request: plug-in support for new devices

### ISSUE THREAD

**Issue by vrv** [tensorflow/tensorflow#4359]
We need to be able to support using new devices in TensorFlow without requiring editing the TensorFlow binary source code.

This requires a few changes.  Among many others:
- The ability to dynamically register an implementation / factories for alternative devices
- New or better APIs for allowing OpKernel's to use arbitrary device resources (e.g., right now we assume either the use of StreamExecutor or EigenDevice implementations, but that's obviously not general).

**Comment by aidan-plenert-macdonald** [comment]
Few questions.

## About Feature Request
By dynamically, I assume you are referring to at compile time, but self-contained code changes. If you mean at runtime via dynamic module loading, I don't know off hand how to do this, but I think we can figure it out.

How do you wish to handle the kernel code? Should I plan to have new hardware supported by developers modifying the `tensorflow/core/kernels/*` and registering the new hardware? Or do you want this also separated. Although this currently is handled with macros and registration systems in the C++, but I ask from a build perspective. We could make a new folder that contains all the code for extra devices and contains all the build scripts over there. It this is done, I can write template files that have all the basic structural code (many projects do this for configuration scripts) that a user could basically rename and be ready to go.

## About TensorFlow

[TensorFlow Device Contexts, Streams, and Context Switching](http://stackoverflow.com/questions/39481453/tensorflow-device-contexts-streams-and-context-switching)

**Comment by vrv** [comment]
Yes, I do mean self-contained code changes: a user should be able to use an existing binary install of TensorFlow that doesn't contain the custom device code, but then call a function to "load a device" and then another function to "load the kernels" for that device.

The latter is already possible for existing devices via the tf.load_op_library() mechanism, so theoretically something similar could be done for a new tf.load_device().

I'll answer your other question on SO, but I don't think the answer there will be instructive for new devices.  Every device has its own execution model, and so the way the device code is written needs to take into account the execution model.  On CPU, the ThreadPoolDevice is just a device that uses Eigen's multi-threaded CPU support to implement operations.  On GPU, we have to manage streams and execution order carefully because GPU devices are asynchronous, not synchronous.

If you told us more about the execution model of your hardware, we might be able to suggest something more relevant -- it's likely that copying the CPU or the GPU way of doing things is not the right solution.

**Comment by aidan-plenert-macdonald** [comment]
I need to talk with my supervisor before giving too many details. [Here](https://www.knupath.com/products/hermosa-processors/) is the public info, but as a high level understanding, the chip is basically a cluster computer reduced to fit on a single chip. The many DSPs need to work together to do anything useful.

I didn't know about the operation interface! Its pretty awesome and I definitely think that is what I want to build.

It would seem that at a minimum, a developer would need to write an Allocator, Device, DeviceFactory, and DeviceContext. This would give a non-functional device because there are no kernels registered to it. As I was developing, I noticed that some kernels seemed to be core functions like ConstOp, AssignOp, NoOp, etc. that are needed for other things to work. It would seem that a user wouldn't want to code these explicitly as they are kind of obvious and redundant. Do you think these can/should be automatically built into the framework so that every device at least has these working out of the box?

**Comment by vrv** [comment]
Yes, those four pieces are the minimal requirements just to get the basics working, and then you'd have to register kernels for every op that was supported.

You're right that some ops are probably trivially implementable as long as some of the basics above are implemented.  We'd have to think about how to 'auto register kernels' for all devices.  However, for things like NoOp, it shouldn't be too hard: it would just be an include and a registration like in https://github.com/tensorflow/tensorflow/blob/5df4c71c86b28c2a4dd746bd67f00fc0281bd24f/tensorflow/core/kernels/no_op.cc#L21

**Comment by aidan-plenert-macdonald** [comment]
How do I add headers to be installed into the TensorFlow include path? And how about for the .proto files? I need to "compile" them to headers and have them installed

I need several framework and common_runtime headers to compile device code. I tried to look through the Bazel files, but couldn't find anything obvious.

**Comment by aidan-plenert-macdonald** [comment]
FYI. I got a fake device working with contexts, factories, kernels and all, so I will begin to try to make it self contained. I posted various questions to StackOverflow and things. Those are more educational. The library problem above is more important for the purposes of this issue.

Thanks a bunch!

**Comment by vrv** [comment]
1) At the moment, unsupported/Eigen/CXX11/Tensor  is required by the framework -- it's how you can map our tensor.h Tensor into typed and shaped tensors.

2) Adding @keveman about how to add TF headers to include path.  I thought https://github.com/tensorflow/tensorflow/blob/master/tensorflow/g3doc/get_started/os_setup.md#setting-up-tensorflow-for-development might be relevant, but I'm not sure.

**Comment by aidan-plenert-macdonald** [comment]
I got all the code down to one file with no changes to the main code base except the below. I list my solution ideas, but they involve modification of the main code so I wanted to make sure its okay.

```
diff --git a/tensorflow/core/util/device_name_utils.cc b/tensorflow/core/util/device_name_utils.cc
index 5816dbd..50309ed 100644
--- a/tensorflow/core/util/device_name_utils.cc
+++ b/tensorflow/core/util/device_name_utils.cc
@@ -162,6 +162,16 @@ bool DeviceNameUtils::ParseFullName(StringPiece fullname, ParsedName* p) {
       }
       progress = true;
     }
+    if (str_util::ConsumePrefix(&fullname, "/kpu:") ||
+        str_util::ConsumePrefix(&fullname, "/KPU:")) {
+      p->has_type = true;
+      p->type = "KPU";  // Treat '/kpu:..' as uppercase '/device:KPU:...'
+      p->has_id = !str_util::ConsumePrefix(&fullname, "*");
+      if (p->has_id && !ConsumeNumber(&fullname, &p->id)) {
+        return false;
+      }
+      progress = true;
+    }
 
     if (!progress) {
       return false;
```
To get rid of this one, I was thinking of using the [existing registration](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/device_factory.cc#L44), but that means exposing this function. I could just make the underlying `std::unordered_map` a private/protected member of the DeviceFactory class and then make DeviceNameUtils a friend. Or just expose the function handle in the header.

```
diff --git a/tensorflow/python/framework/device.py b/tensorflow/python/framework/device.py
index 8f5125d..505c2a9 100644
--- a/tensorflow/python/framework/device.py
+++ b/tensorflow/python/framework/device.py
@@ -155,7 +155,9 @@ class DeviceSpec(object):
         elif ly == 2 and y[0] == "task":
           self.task = y[1]
         elif ((ly == 1 or ly == 2) and
-              ((y[0].upper() == "GPU") or (y[0].upper() == "CPU"))):
+              ((y[0].upper() == "GPU") or 
+               (y[0].upper() == "CPU") or
+               (y[0].upper() == "KPU"))):
           if self.device_type is not None:
             raise ValueError("Cannot specify multiple device types: %s" % spec)
           self.device_type = y[0].upper()
```
If there is Python access to the same interface as above, then I can just use that. I am a fan of less configuration.

**Comment by vrv** [comment]
I think if you name your device "/device:KPU:0" instead of just "/KPU:0", it should work without any additional edits.  "/cpu:0" and "/gpu:0" were shortcuts before we realized it was helpful to have the "/device" qualifier like we do for the other parts of a full device name.  Can you give that a try and let me know?

**Comment by aidan-plenert-macdonald** [comment]
Yup that worked!

```
Device mapping:
/job:localhost/replica:0/task:0/device:FPU:0 -> physical description
/job:localhost/replica:0/task:0/device:FPU:1 -> physical description
/job:localhost/replica:0/task:0/device:FPU:2 -> physical description
/job:localhost/replica:0/task:0/device:FPU:3 -> physical description
```

Once I figure out how to get those includes and compile it outside of TF, I will submit a pull request within documentation a la ["Adding an Op"](https://www.tensorflow.org/versions/r0.10/how_tos/adding_an_op/index.html)

After that. I would like to discuss how to best support my particular hardware. My company probably wants this to be private as I will need to tell you details, so would you be willing to do it over email and not as a GitHub Issue?

**Comment by vrv** [comment]
Adding @petewarden

**Comment by aidan-plenert-macdonald** [comment]
@keveman 

I have been manually adding headers to my include path and I need,

```
tensorflow/core/framework/device_attributes.pb_text.h
tensorflow/core/framework/op_segment.h
tensorflow/core/framework/resource_mgr.h
tensorflow/core/framework/tensor.pb_text.h
tensorflow/core/framework/type_index.h
tensorflow/core/graph/edgeset.h
tensorflow/core/graph/graph.h
tensorflow/core/graph/types.h
tensorflow/core/arena.h
tensorflow/core/lib/gtl/int_type.h
tensorflow/core/lib/gtl/iterator_range.h
tensorflow/core/public/session_options.h
```

Manually adding symlinks to for all those above headers tothe TensorFlow source, I still get this error,
```
g++ -std=c++11 -shared fpu_device.cc -o fpu.so -fPIC -I /usr/local/lib/python2.7/dist-packages/tensorflow/include
In file included from /usr/local/lib/python2.7/dist-packages/tensorflow/include/tensorflow/core/common_runtime/local_device.h:19:0,
                 from fpu_device.cc:10:
/usr/local/lib/python2.7/dist-packages/tensorflow/include/tensorflow/core/common_runtime/device.h: In member function ‘std::string tensorflow::Device::DebugString() const’:
/usr/local/lib/python2.7/dist-packages/tensorflow/include/tensorflow/core/common_runtime/device.h:139:74: error: ‘ProtoDebugString’ was not declared in this scope
   string DebugString() const { return ProtoDebugString(device_attributes_); }
                                                                          ^
fpu_device.cc: In member function ‘virtual tensorflow::Status tensorflow::FPUDevice::MakeTensorFromProto(const tensorflow::TensorProto&, tensorflow::AllocatorAttributes, tensorflow::Tensor*)’:
fpu_device.cc:84:41: error: ‘ProtoDebugString’ was not declared in this scope
            ProtoDebugString(tensor_proto));
```

**Comment by petewarden** [comment]
ProtoDebugString() is generated by tensorflow/tools/proto_text, which is a bit like protoc but for a minimal footprint class instead of a full protobuf one.

**Comment by aidan-plenert-macdonald** [comment]
Thanks. Do you know how I can get it to be recognized when I am compiling it into an external module?

I am currently running 
```
g++ -std=c++11 -shared fpu_device.cc -o fpu.so -fPIC -I /usr/local/lib/python2.7/dist-packages/tensorflow/include
```

**Comment by DavidNorman** [comment]
Hello.  I think I should introduce myself.  I am working for GraphCore, a UK startup making a graph processor / accelerator.  It seems that I am traveling the same path as Aidan.

I have a vaguely functioning device, although I was stuck on the 'having to edit the device_name_utils.cc and device.py' problem.   It would be a very useful thing to be able to build a dynamically linked device/kernels module indeed.  I have isolated all of my device/kernel code into the third_party branch.  Does this make sense?

My device depends on an external library, and so I have also added some code to the workspace.bzl file.  With only one workspace for the whole of tensorflow, there doesn't seem to be a way around this.  Is there is something that I have missed about bazel that would allow me to break the need for that?

Cheers

**Comment by DavidNorman** [comment]
I saw the slashdot conversation about [DeviceContexts](http://stackoverflow.com/questions/39399890/tensorflow-device-contexts?noredirect=1&lq=1).

Can you clarify the difference between Devices and DeviceContexts.  Why would I not store device specific information in my Device class?  For instance, handles to device specific structures, mapped memory, etc?

**Comment by aidan-plenert-macdonald** [comment]
@DavidNorman If you look back a [couple posts ago](https://github.com/tensorflow/tensorflow/issues/4359#issuecomment-247488042) I had the same problem. The trick is calling you device something like ["device:FPU"](https://github.com/tensorflow/tensorflow/issues/4359#issuecomment-247488613). In my pull request, [I do this](https://github.com/tensorflow/tensorflow/pull/4520/files#diff-7c3aa428ac49c5f21774505a5985d051R103)

Also, @DavidNorman were you able to get your code to compile as a .so file? I have trouble with the includes. Or are you compiling into the TensorFlow main body

**Comment by aidan-plenert-macdonald** [comment]
@DavidNorman with regards to the Device Contexts, I forget were I heard it, but I think @vrv told me that GPUs have multiple contexts which handle different parts of computation like memory access, pushing kernels onto the device, etc.

Also, if you could put both your questions on StackOverflow, I will answer them there for better accessibility in the future. eg. [My question about compilation](http://stackoverflow.com/questions/39643820/adding-more-files-to-the-tensorflow-include-path)

**Comment by DavidNorman** [comment]
@aidan-plenert-macdonald thanks for the info  I do not have a separate .so file at the moment (or a .dylib as it would be on my Mac).   I have a few more hurdles to get past before I try to make that happen i think.  The requirement for an external library means I have to modify the workspace.bzl, so I'm not sure that there is too much point unless I can figure out how to avoid that.  

Have posted [this question](http://stackoverflow.com/questions/39656677/tensorflow-device-vs-devicecontext) on slashdot

**Comment by aidan-plenert-macdonald** [comment]
@vrv I was wondering if you have any further information about adding includes. We just need those to be done.

**Comment by vrv** [comment]
I really don't know :(  Going to assign this one to @keveman (I'll be OOO soon for a week).

**Comment by aidan-plenert-macdonald** [comment]
Got it!! Simple one liner in a BUILD file.

[Bottom of the page](https://github.com/tensorflow/tensorflow/pull/4520/files?diff=unified)

**Comment by aidan-plenert-macdonald** [comment]
@DavidNorman 
You were looking at the difference between DeviceContexts and Devices. I am debating between having Device represent actual hardware with several DeviceContext's that control virtual compute nodes (my device can allocate compute resources dynamically) which would each handle a vertex in my TF compute graph (I will use node to indicate hardware and vertex for TF operations). My concern is that TF doesn't appear to support binding Allocators to DeviceContexts which would be needed so I can make sure allocation for each TF vertex happens within my virtual compute node. On the other hand if I have Devices be my virtual compute nodes and bind each one to a TF vertex, then I can't dynamically size them as I believe the creation of devices in the DeviceFactories happens before the TF compute graph is assembled.

Can I ask how you are assigning compute resources to nodes in the TF compute graph? Are you using Device to control hardware or is this done with contexts? Do you have similar problems with allocation? Do you know anything about TF's ability to do automatic resource allocation? I believe TF can auto-assign devices to operations.

@vrv Is there any way to bind Allocators to specific device contexts?

**Comment by vrv** [comment]
@aidan-plenert-macdonald: Typically an Allocator manages the memory for an entire device.

In the GPU case, there is one allocator for each GPU device, but a GPU device has hundreds to thousands of "cores", each which has access to the global memory that the allocator is responsible for allocating, and programs are responsible for setting properly.  CUDA has ways in its programming model to additionally have local fine-grained sharing of memory among cores, and that is specific to the cuda programming language, and we don't touch that really, since it's not part of the global memory pool.

I suspect that your device is really more like many loosely coupled cores with message passing, which is a model we haven't really optimized for.  We would normally treat each 'core' as a separate device, and then your memcopy primitives (device to device) would use whatever high-performance inter-node communication primitives you wanted.

Alternatively, if you'd rather treat your entire processor as a single "device", it might still be possible: allocator.h has an AllocatorAttributes structure, which is a 64-bit opaque value blob.  The top 8 bits of that structure we've made device specific.

OpKernel has a GetAllocator() function that takes allocator attributes, so it might be possible for you to have the DeviceContext contain information that an OpKernel can use to set the appropriate bits of the allocator, and then you'd implement DeviceBase::GetAllocator() to return a different allocator based on the top 8 bits of the AllocatorAttributes.  If you used all 8 bits as an allocator id, you'd be able to address 256 different allocators in a single device.

Without knowing too much about your device, I'm not sure which approach is better, but those might help you make progress.

**Comment by aidan-plenert-macdonald** [comment]
@vrv I like the 8 bit option. I may run into problems with the 256 limit later, but I think for now this is a really good option.

What would I have to change to make sure that those 8 bits get set appropriately? I tracked GetAllocator to the Executor class where it appears that I could modify the AllocatorAttributes and the Executor has virtual functions which make it appear as though I could replace it, but I don't see any registration for the ExecutorImpl.

Now if I were to just have a TF Device per virtual compute node with a single allocator each. I would like to have a couple device numbers/config to specify the physical hardware and then the virtual compute nodes within it. the Distribute TF seems to use more numbers,

```python
tf.train.ClusterSpec({
    "worker": [
        "worker0.example.com:2222", 
        "worker1.example.com:2222",
        "worker2.example.com:2222"
    ],
    "ps": [
        "ps0.example.com:2222",
        "ps1.example.com:2222"
    ]})
```

Internal to my device, I could treat it as a cluster computer. Is the distributed computing very extensible so I could rework the internals?

**Comment by vrv** [comment]
We could change the 8 bits to a larger number of bits before 1.0 I guess.  Or perhaps one day maybe extend the value to be larger than 32 bits in a backwards compatible way.

The OpKernel::allocate_output() function can set the bits of the AllocatorAttributes however it wants.  I was thinking, but not sure if it works off hand:

```
FPUDeviceContext* fdc = static_cast<FPUDeviceContext*>(context->device_context());
int32 device_id = fdc->device_id();
AllocatorAttribute attr;
attr.value |= (device_id && 0x000000FF) << 24; 

Tensor* output;
context->allocate_output(0, shape, &output, attr);
```

Or something like that.

**Comment by aidan-plenert-macdonald** [comment]
@vrv Do you know the answer to [this](http://stackoverflow.com/questions/39797095/tensorflow-custom-allocator-and-accessing-data-from-tensor)?

**Comment by vrv** [comment]
Sorry, was on vacation for the past week -- just answered.

**Comment by aidan-plenert-macdonald** [comment]
I wrote a TensorFlow framework for a new device. I then wrote a Matix Multiplication that does nothing but print something when the constructor and Compute functions are called. The code looks like,

    #include "tensorflow/core/framework/op.h"
    #include "tensorflow/core/framework/op_kernel.h"
    #include "tensorflow/core/common_runtime/dma_helper.h"
    
    namespace tensorflow {
    
      class MatMulOp : public OpKernel {
      public:
        explicit MatMulOp(OpKernelConstruction* context) : OpKernel(context) {
          KPU_DEBUG("Init");
        }
    
        void Compute(OpKernelContext* context) override {
          KPU_DEBUG("Compute");
        }
      };
    
      REGISTER_KERNEL_BUILDER(Name("MatMul").Device(DEVICE_KPU), MatMulOp); 
    }

Unfortunately, this code isn't being run during a simple test program taken directly from TensorFlow,

    import tensorflow as tf
    mod = tf.load_op_library('kpu.so')
    with tf.device('/device:KPU:0'):
      a = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[2, 3], name='a')
      b = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[3, 2], name='b')
      c = tf.matmul(a, b)
      # Creates a session with log_device_placement set to True.
    
    sess = tf.Session(config=tf.ConfigProto(log_device_placement=True))
    # Runs the op.
    print sess.run(c)

Oddly enough, the NoOp code is being run (I remove the print statements, but it was printing things out),
```
namespace tensorflow {

  class NoOp : public OpKernel {
  public:
    explicit NoOp(OpKernelConstruction* context) : OpKernel(context) { }
    void Compute(OpKernelContext* context) override { }
    bool IsExpensive() override { return false; }
  };

  REGISTER_KERNEL_BUILDER(Name("NoOp").Device(DEVICE_KPU), NoOp);

}
```

**Comment by vrv** [comment]
I think in this case, our optimizations for constant folding are foiling you :).   (Because c can be computed statically, we pre-compute it on the first run of the graph on CPU, and then there's nothing to run on your device).

1) You can disable this by disabling optimizations in ConfigProto.GraphOptions.

2) You can redefine your graph to have the matmul take in two 'variables', in which case it's not constant foldable anymore.

**Comment by aidan-plenert-macdonald** [comment]
I think I will make them variables because I need to do that anyway. How do I fix,
```
tensorflow.python.framework.errors.InvalidArgumentError: Cannot assign a device to node 'b': Could not satisfy explicit device specification '/device:KPU:0' because no supported kernel for KPU devices is available.
Colocation Debug Info:
Colocation group had the following types and devices: 
Identity: CPU 
Assign: CPU 
Variable: KPU CPU 
	 [[Node: b = Variable[container="", dtype=DT_FLOAT, shape=[10,10], shared_name="", _device="/device:KPU:0"]()]]
 
```

**Comment by vrv** [comment]
You have to register kernels for a lot of standard ops on your KPU device.

(NoOp, Variable, Assign, Identity, etc).  I think the generic implementations are registerable via header files like variable_op.h etc -- look for the GPU kernel registrations to get a sense of what you probably want to do for your KPU.  It's likely that most of these do not require a custom implementation on your device.

**Comment by aidan-plenert-macdonald** [comment]
Perhaps I should rephrase my question. I have kind of been shooting in the dark when registering the kernels. Is there a way to get the exact name of the kernel I am missing so I can implement it over?

**Comment by vrv** [comment]
```
Identity: CPU 
Assign: CPU 
Variable: KPU CPU 
```

This is saying that these three groups of nodes needed to colocated on a single device, but the only common device implementation was CPU, so these nodes could not be placed on KPU.

If you register "Identity" and "Assign" on KPU, these three nodes could then be colocated together on KPU.

**Comment by aidan-plenert-macdonald** [comment]
The master branch of TF recently switched to Protobuf 3.1.0 and my companies drivers use 3.0.0. I am now having trouble compiling. is there a good clean way to make backward compatibility work out? Can I force a Protobuf roll back with TF?

**Comment by vrv** [comment]
You could locally change the workspace file to point to the earlier version of protobuf, but I suspect we moved to 3.1.0 for a good reason, so it's not something we could accept.  Might also be worth asking the protobuf team at github.com/google/protobuf ...

**Comment by aidan-plenert-macdonald** [comment]
My team fixed it on our side. Now I have reason to believe that TF is messing with the pointers I am returning from my custom allocator which are actually pointers to structures that control the hardware. It seems that the pointers are getting used within [Buffer](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/tensor.cc#L87) and then the Buffer tries to free them as if they were a float pointer which causes a SegFault. Is this something that might happen? How can I fix it?

**Comment by aidan-plenert-macdonald** [comment]
@vrv 

I am getting segfaults from the destructors where the program terminates (below is a stacktrace).
```
Program received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) backtrace
#0  0x0000000000000000 in ?? ()
#1  0x00007fffe0524e94 in tensorflow::(anonymous namespace)::Buffer<float>::~Buffer() () from /usr/local/lib/python2.7/dist-packages/tensorflow/python/_pywrap_tensorflow.so
#2  0x00007fffe0524f41 in tensorflow::(anonymous namespace)::Buffer<float>::~Buffer() () from /usr/local/lib/python2.7/dist-packages/tensorflow/python/_pywrap_tensorflow.so
#3  0x00007fffe0525fd6 in tensorflow::Tensor::~Tensor() () from /usr/local/lib/python2.7/dist-packages/tensorflow/python/_pywrap_tensorflow.so
#4  0x00007fffe0162514 in tensorflow::VariableOp::Var::~Var() () from /usr/local/lib/python2.7/dist-packages/tensorflow/python/_pywrap_tensorflow.so
#5  0x00007fffe051a474 in tensorflow::ResourceMgr::Clear() () from /usr/local/lib/python2.7/dist-packages/tensorflow/python/_pywrap_tensorflow.so
#6  0x00007fffe051a501 in tensorflow::ResourceMgr::~ResourceMgr() () from /usr/local/lib/python2.7/dist-packages/tensorflow/python/_pywrap_tensorflow.so
#7  0x00007fffe04505a0 in tensorflow::Device::~Device() () from /usr/local/lib/python2.7/dist-packages/tensorflow/python/_pywrap_tensorflow.so
#8  0x00007fffe894e6d1 in tensorflow::KPUDevice::~KPUDevice (this=0xe65ee0, __in_chrg=<optimized out>) at kpu/device.cc:49
#9  0x00007fffe894e700 in tensorflow::KPUDevice::~KPUDevice (this=0xe65ee0, __in_chrg=<optimized out>) at kpu/device.cc:52
#10 0x00007fffe045284e in tensorflow::DeviceMgr::~DeviceMgr() () from /usr/local/lib/python2.7/dist-packages/tensorflow/python/_pywrap_tensorflow.so
#11 0x00007fffe035b087 in tensorflow::DirectSession::~DirectSession() () from /usr/local/lib/python2.7/dist-packages/tensorflow/python/_pywrap_tensorflow.so
#12 0x00007fffe035b1d1 in tensorflow::DirectSession::~DirectSession() () from /usr/local/lib/python2.7/dist-packages/tensorflow/python/_pywrap_tensorflow.so
#13 0x00007fffe04401b7 in TF_DeleteSession () from /usr/local/lib/python2.7/dist-packages/tensorflow/python/_pywrap_tensorflow.so
#14 0x00007fffdf16a90b in _wrap_TF_DeleteSession () from /usr/local/lib/python2.7/dist-packages/tensorflow/python/_pywrap_tensorflow.so
#15 0x000000000049968d in PyEval_EvalFrameEx ()
```

I am using the default VariableOp by just importing the header and registering it to my device. I am unsure what I am doing wrong that would cause it to segfault. I believe it is trying to free the pointers to my tensor data incorrectly, but I am neither sure of that, nor is it clear how to fix it.
```c++
#include "kpu/kpu.h"
#include "tensorflow/core/kernels/variable_ops.h"

#include "tensorflow/core/framework/op_kernel.h"
#include "tensorflow/core/framework/register_types.h"
#include "tensorflow/core/lib/core/errors.h"
#include "tensorflow/core/platform/types.h"

namespace tensorflow {

#define REGISTER_KPU_KERNELS(type)                                       \
  REGISTER_KERNEL_BUILDER(                                               \
      Name("Variable").Device(DEVICE_KPU).TypeConstraint<type>("dtype"), \
      VariableOp);                                                       \
  REGISTER_KERNEL_BUILDER(Name("TemporaryVariable")                      \
                              .Device(DEVICE_KPU)                        \
                              .TypeConstraint<type>("dtype"),            \
                          TemporaryVariableOp);                          \
  REGISTER_KERNEL_BUILDER(Name("DestroyTemporaryVariable")               \
                              .Device(DEVICE_KPU)                        \
                              .TypeConstraint<type>("T"),                \
                          DestroyTemporaryVariableOp);                   \
  REGISTER_KERNEL_BUILDER(Name("IsVariableInitialized")                  \
                              .Device(DEVICE_KPU)                        \
                              .TypeConstraint<type>("dtype")             \
                              .HostMemory("is_initialized"),             \
                          IsVariableInitializedOp);

  TF_CALL_GPU_NUMBER_TYPES(REGISTER_KPU_KERNELS);

}  // namespace tensorflow 
```

Could you help me understand what is happening?

**Comment by vrv** [comment]
I'm not entirely sure without being able to see the code, unfortunately.  The destructor of Tensors should just call the Deallocate method of your custom Allocator -- you might be able to trace that code to figure out what's going on?

**Comment by aidan-plenert-macdonald** [comment]
Fixed it. I was calling `delete kpu_allocator_` in my Device destructor while I think TF still had reference pointers floating around. Took way too long to solve.

**Comment by vrv** [comment]
@aidan-plenert-macdonald do you have any interesting updates so far?

**Comment by aidan-plenert-macdonald** [comment]
I have been working more on coding for my hardware. I have the TensorFlow stuff working. I just want to get my hardware working with TF and then make a couple more changes to my TF code to clean it up a bit more. Once I get that all cleaned up, I can start working on the documentation for this. I just want to work out all the bugs and oddities first.

**Comment by aidan-plenert-macdonald** [comment]
@vrv I got things working and am at a good rebase/pausing point.

Everything works without changing TF beyond installing more header files (simple couple lines in Bazel). Let me know what goals you have on your side. I can write documentation, examples, or just be a resource for other people trying to extend TF to new hardware. A couple pieces of interesting hardware are becoming popular/relevant: Intel Phi's, TrueNorth, SyNAPSE, etc. If people want to add these to TF, then I could help them get over the learning curve, as a community that knows how to get this done is probably better than documentation.

**Comment by vrv** [comment]
Awesome, we're really glad to hear that!

There are a couple of audiences that would be interested in general, and you are right that having documentation might go stale right now.  There's also [XLA](https://github.com/tensorflow/tensorflow/blob/47ef3c5eca776a2d10249f987cca406577389db5/tensorflow/g3doc/resources/xla_prerelease.md) which is relevant to some but not all hardware developers.

Mostly, I just wanted validation that we didn't have to add code to core TF to add new devices, and it seems like (pending your PR to add those missing files), it should work.  We'd be happy to have documentation or an example to show the minimal set of things that device implementors might need to think about when adding new devices.

Maybe it makes sense to start out with a blog post or something describing what you had to do, (so it's discoverable), and then we can see what makes sense to have either on GitHub or in the codebase?

**Comment by RichDubielzig** [comment]
Hi Vijay,

I'm taking over for Aidan in adapting TensorFlow for the Knureon.  I am currently working on some automated scripts that will help our work in sync with the tip of development on TensorFlow.

I am currently trying to figure out how to get my local TensorFlow build to run validation tests to so that I can be confident that the code snapshot I pulled from github is consistent with itself before merging it into our own work.  Unfortunately Bazel is still a little new to me, so I am having trouble understanding the calls I need to make to get Bazel to run all tests in a class (we want to exclude the GPU tests for now), and not simply run line-by-line tests as described in your "Adding a New Op" tutorial.  Is there a syntax similar to LLVM's 'lit' tool that allows for a class or regexp of test names to be run?

-Rich

**Comment by vrv** [comment]
I don't know llvm's lit tool, but you can do
```
bazel test //tensorflow/python/... 
```
to build all targets under //tensorflow/python

If you don't run with --config=cuda, you will not run any tests that require a GPU (our code *should* automatically skip tests that require a GPU to run).

If you want to run a specific target, you can do:
```
bazel test //tensorflow/python:constant_op_test
```
or something like that.

**Comment by llhe** [comment]
Hi @vrv , will the current way (mainly writing customized kernels) of adding new devices changed after the introduce of XLA compiler?

**Comment by llhe** [comment]
And more specifically, will XLA also target for mobile devices? E.g. with customized DNN acceleration chips.

**Comment by vrv** [comment]
@prb12 @michaelisard can you comment more? I wouldn't say writing a device today is that simple, considering there are 300+ ops to implement.  XLA should make that job a lot easier.

**Comment by prb12** [comment]
I think this depends a lot on the capabilities of the hardware device you are planning to support.
There are several scenarios which I can imagine:

- If the hardware is a 'one-trick-pony' accelerator (e.g. it provides a small number of fixed functions) then it may be sensible to expose those from a traditional TF device.  e.g. register new device-specific kernels for existing ops such as MatMul, or if the hardware does some very strange things then add custom ops and ideally some sort of GraphOptimizationPass which uses device-specific knowledge to rewrite existing TF programs to use your ops where it makes sense.

- The above approach probably works quite well if you're wanting to add some sort of black-box library implementing eg. something like MKL.  You don't want to break portability of people's TF programs, but you want to rewrite them to use custom low-level functionality at runtime.

- For an accelerator which provides a powerful set of generic computation support it is much easier to implement the small number of primitives required by XLA than the many 100s of ops registered in TensorFlow.  The TF2XLA bridge automatically identifies parts of graphs to hand to XLA and converts the TF ops into XLA primitives.  The set of ops which are currently translatable is enough to handle complex models like inception.    

- Implementing XLA support for a processor supported by LLVM is not very difficult.  For something stranger you will need to write your own XLA backend.  This is strictly easier than implementing the equivalent TF ops, but you will need some compiler expertise.

**Comment by RichDubielzig** [comment]
We are hoping to implement as many TF operations as we can for the Knureon.  To the extent we have a 'trick' we are focusing on sparse matrices because we believe we can show a significant performance improvement.

Is there a white paper available on XLA?  Is the source on a branch?  I don't see sparse matrices mentioned in  the prerelease.

Also, is there a target release date for XLA?  Our own schedule may require us to support traditional TF kernels for our first release, at least.

**Comment by llhe** [comment]
@prb12 Thanks for the explanations, and it makes my understanding much clear.

**Comment by RichDubielzig** [comment]
Thank you all for your help.  I asked this question to Vijay in an email, but I think I'd like to post it here too.

---------

I've been trying to understand how backpropagation is implemented in tensorflow models.  I'm just asking you to let me know if I have any major misunderstandings about the nature of TensorFlow here.  Thank you again for your efforts to help me out.

I'm basing this interpretation on the graph generated when I add a FileWriter to the mnist_softmax.py tutorial example (attached here, also with a relu activation added as an experiment).

It looks like the act of placing an optimizer such as a GradientDescentOptimizer into a computation flow creates a new subgraph that mirrors the original graph, except that
1. Data flows through the nodes in reverse order and
2. The nodes themselves are replaced with their gradients.

One training run, therefore will consist of a forward pass through the computation steps as described in the python file, culminating in the loss calculation.  If we are using minibatches, this is done a number of times.  Then, a second pass backwards through the gradients of cross entropy will cause the Variable objects to be updated with trained values and the run() command will complete.

Some questions:
1.  I see that I can replace the "relu" activation with "atan" and have the model run without crashing.  Why does this work?  I don't see any atan gradient function under the kernels/ directory.
2.  During the course of a run() command, do the weight & bias variables really get overwritten by backpropagation, or does TF keep around the original values and store the adjusted ones in new reserved memory?
3.  GradientDescentOptimizer() is called with a "learning rate" '0.5'.  Is this argument only used to scale the original loss function, or is it passed through all the way back through each layer of the network?


[mnist_softmax.py.zip](https://github.com/tensorflow/tensorflow/files/671609/mnist_softmax.py.zip)

**Comment by yaroslavvb** [comment]
It helps to think of different layers of TensorFlow -- there's the client
side which creates computational graph, and TensorFlow runtime side which
executes it.

`tf.gradients([a],[b])` is a "client-side" operation that adds nodes to the
graph that give gradient of a with respect to b. In a sense it's not part
of core TensorFlow since it's not accessible to people using TensorFlow
from other languages. Also, we've been finding that graph created by
`tf.gradients` is memory inefficient, and a modified version is needed to
make some models fit.

A visualization of a typical graph you get from tf.gradients is attached as
picture 1. Nodes b_i compute left product of a vector x with jacobian
evaluated at a_i, so you can see that each node has two inputs. This is the
core step of "reverse-mode automatic differentiation." This can be a fused
implementation corresponding to a single TensorFlow node, or it can be
implemented as a TensorFlow subgraph. For instance, gradient of tf.square
is implemented as tf.mul(2, x) rather than a dedicated SquareGrad.

From graph you can see that to compute it you need enough memory to store
a1/a2/a3. If that's too much, one could save memory by using a modified
implementation in attachment 2.

Regarding questions

1. You can see the graph you get as follows

a = tf.placeholder(tf.float32)
b = tf.atan(a)
tf.gradients([b],[a])
print(tf.get_default_graph().as_graph_def())

It seems to be implementing atanh(z) grad using 1/(1-z^2) formula.

2. ApplyGradientDescent(var, alpha, delta) will overwrite the value of var
with new weight values.

3. GradientDescentOptimizer translates into ApplyGradientDescent(var, 0.5,
gradient) , so 0.5 is only introduced at the last step

**Comment by yaroslavvb** [comment]
@RichDubielzig  It helps to think of different layers of TensorFlow -- there's the client side which creates computational graph, and TensorFlow runtime side which executes it.

`tf.gradients([a],[b])` is a "client-side" operation that adds nodes to the graph that give gradient of a with respect to b. In a sense it's not part of core TensorFlow since it's not accessible to people using TensorFlow from other languages. Also, we've been finding that graph created by `tf.gradients` is a memory hog, and a modified version is sometimes needed to make models fit.

Here's what a typical graph you get from tf.gradients looks like
![ad-graphs1](https://cloud.githubusercontent.com/assets/23068/21486407/f172bc40-cb68-11e6-9b47-f87e85a99293.png)

Nodes `b_i` compute left product of a backprop vector with jacobian evaluated at `a_i`, so you can see that each node has two inputs. This is the core step of "reverse-mode automatic differentiation." This can be a fused implementation corresponding to a single TensorFlow node, or it can be implemented as a TensorFlow subgraph. For instance, `grad(a,b)` of `tf.square` is implemented as `a*2*b` rather than a fused `SquareGrad` op.

From graph you can see that to compute it you need enough memory to store `a1/a2/a3`. If that's too much, one could use a modified implementation below which saves memory at the cost of recomputing `a2`

![ad-graphs2](https://cloud.githubusercontent.com/assets/23068/21486412/0c68954c-cb69-11e6-8467-22c402c529c2.png)

So the typical implementation of training that client constructs gradient graph using `tf.gradients` or related, and updates weight variable using `var.assign_add(lr*grad)` or a fused version that comes of `apply_gradient` method

Regarding questions

1. You can see the graph you get as follows

```
a = tf.placeholder(tf.float32)
b = tf.atan(a)
tf.gradients([b],[a])
print(tf.get_default_graph().as_graph_def())

```
It seems to be implementing `atanh(z)` grad using `1/(1-z^2)` formula.

2. `ApplyGradientDescent(var, alpha, delta)` will overwrite the value of var
with new weight values.

3. GradientDescentOptimizer use translates to `ApplyGradientDescent(var, 0.5, gradient)` , so `0.5` is only introduced at the last step

**Comment by RichDubielzig** [comment]
Thank you Yaroslav for your answers, this was very helpful.

**Comment by RichDubielzig** [comment]
Hello,

I am currently working my way through understanding the backend of the XAL framework.  I currently have two questions:

1.  How do you debug the TensorFlow library?  For example, I wanted to try breaking on this debug message:

`I tensorflow/compiler/xla/service/service.cc:187]   StreamExecutor device (0): <undefined>, <undefined>`

I tried 
```
gdb --args python3 mnist_softmax_xla.py
set follow-fork-mode child
b service.cc:187
r
```

But the breakpoint was skipped over.

2.  How does the JIT compiler "link in" on-core runtime libraries?  In order to take advantage of XAL, we'd need to have some point at which we link in a bit of code to handle general setup and teardown of a core, as well as communication back to the host.  We also have some TF-specific memory-handling code that would need to go in as well.  At what point do these libraries get added to the HLO? (We can provide the libraries in IR format)

**Comment by RichDubielzig** [comment]
I *think* I've answered my second question.  Looking at compiler/xla/service/gpu/llvm_gpu_backend, I see the function LinkLibdeviceIfNecessary, which takes a directory path as a direct argument and the bitcode filename to link from a configuration argument.  It then does a bitcode-level linkage (which is our preference) against the runtime library and passes the result back to CompileModuleToPtx.

So it looks like our simplest approach is to create a .bc runtime library and reuse the GPU compilation code to compile to our own ISA.

A follow-up question: I don't see anything in the GPU linker or compiler that checks for maximum code size.  Is there any mechanism to guarantee or check that a requested compilation block will return a kernel under a given number of instructions?

**Comment by aidan-plenert-macdonald** [comment]
For question 1 http://stackoverflow.com/questions/33746071/tensorflow-core-debug-missing-debug-symbols might help

**Comment by RichDubielzig** [comment]
That looked promising, but apparently the built tensorflow object is the same as what got installed by pip.  

```
rdubielzig@swpl-000224:~/.cache/bazel/_bazel_rdubielzig/460f4b3af716bd4767655e290c45766f/execroot/tensorflow-knureon/bazel-out/local-py3-dbg/bin/tensorflow/python$ md5sum ~/.local/lib/python3.5/site-packages/tensorflow/python/_pywrap_tensorflow.so 
982e011106ed013c21638157659b34e5  /home/rdubielzig/.local/lib/python3.5/site-packages/tensorflow/python/_pywrap_tensorflow.so
rdubielzig@swpl-000224:~/.cache/bazel/_bazel_rdubielzig/460f4b3af716bd4767655e290c45766f/execroot/tensorflow-knureon/bazel-out/local-py3-dbg/bin/tensorflow/python$ md5sum _pywrap_tensorflow.so
982e011106ed013c21638157659b34e5  _pywrap_tensorflow.so
```

**Comment by vrv** [comment]
@RichDubielzig perhaps @eliben might be able to help or point you at somebody who can help regarding your 'maximum code size' question.

**Comment by eliben** [comment]
@RichDubielzig not sure I understand the maximum code size requirement. Care to elaborate? Where in the API would you expect to see this?

**Comment by RichDubielzig** [comment]
I would expect to see code size considerations taken in generation of kernel code from LLVM IR.  The instuction memory for individual cores in Knureon is quite small, and it looks like XLA could JIT kernel code  that would overrun program memory.  Is there a mechanism to break up a block of IR into smaller pieces?

**Comment by eliben** [comment]
@RichDubielzig this functionality is not supported by the current CPU/GPU backends; however if you'd plan to add a custom backend for your architecture you could implement this.

**Comment by aidan-plenert-macdonald** [comment]
@eliben Do you mean implement it in the main TensorFlow, or just in our library? Can you recommend a good place that it might fit into nicely?

**Comment by RichDubielzig** [comment]
I'm afraid I'm having trouble understanding how to handle synchronization and asynchronous kernel resources when developing in TensorFlow using TensorFlow objects.  Here is my problem:

The knureon system can be thought of as a large pool of completely independent cores which can be assigned operations at any time.  So, for example, given the data flow below:

A: m = a * b
B: n = c *d
C: o = e * f

If I have 128 available cores, then I might request A to run on 64 cores, B to run on another 64 cores, and then I would need to wait for one of the first two operations to complete before I can C on 64 cores.  To this end, I have created a little launch queue object that can hold computation requests until compute resources are available, then run.  In pseudocode, my naive OpKernel Compute() implementation is below.  Note that this employes a Knureon-specific command queue object which can be used both to launch operations and to pend on running ones.  Multiple queues are allowed to exist and run in parallel on a single Knureon system.

```
Compute( context  input):
  Get input tensors A & B from context
  Also get a Knureon command queue handle Q from the DeviceContext object.
  Figure out the dimensions of the tensors
  Assemble a call f(A,B,dims) to my Knureon matrix multipllier algorithm
  Queue the tuple (f(),Q) in the launch queue.
  We can now block on Q until f() completes.

(in launch queue, in a separate thread:)
  Wait for resources to run f()
  dispatch f() using the queue Q on the Knureon complex.
```

The problem should be apparent:  I don't want TensorFlow to sit around waiting on Q if there are other operations that can be deployed right away on available resources.  The alternative *appears* to be to use an asynchronous opkernel:

```
ComputeAsync( context  input, DoneCallback done):
  Get input tensors A & B from context
  Also get a Knureon command queue handle Q from the DeviceContext object.
  Figure out the dimensions of the tensors
  Assemble a call f(A,B,dims) to my Knureon matrix multipllier algorithm
  Queue the tuple (f(), Q, done) in the launch queue.
  Return immediately

(in launch queue, in a separate thread:)
  Wait for resources to run f()
  run f() using Q on the Knureon complex and spawn a task that will wait on Q.  This task will call the ComputeAsync done() call back when f() is complete.
```
But I am not sure if this is the right approach, as it seems to be reserved for high-latency operations such as receiving over a network.  I've been over and through the code, and unfortunately this has only increased my confusion.  I see that the GPU uses a StreamExecutor, but the presence of all the ThenXxxx() functions in the Stream prototype makes me suspect that it is not what I want.

I have also noticed that OpKernel Compute() methods can be called in parallel from concurrent Executor threads.  So do I even need to sweat about parallel execution at all? When an OpKernel's Compute() method is invoked, am I guaranteed that there will be no benefit to running asynchronously because all the other OpKernels managed by my threads' Executor have data dependencies on my operation?

Thank you in advance and my apologies for rambling.  I've had to spend a few days figuring out how to phrase this question in a coherent manner, and I'm not sure I've met my goal, so if you need anything clarified please let me know.

**Comment by CUinNYC** [comment]
Hi! As many other I am really interested in defining my own devices. 

I read this thread as well as some others. Is there any documentation available to implement new devices (either in the TensorFlow source code or as a separate module)?

@RichDubielzig  and @aidan-plenert-macdonald  were working on a guide, but at the moment I could only see this:
https://github.com/knuedge/tensorflow/blob/36e0cdf04f294bfd51931d4f78e291590ed0d3ec/tensorflow/g3doc/hardware/adding_support/index.md

In there anything more recent (targeting the release 1.0)?

Thank you!

**Comment by vrv** [comment]
@CUinNYC indeed, that doc is a great attempt to show it's done, and that's the basic idea of how the scaffolding works, but the implementation details for every new hardware device means it takes care to figure out exactly how to implement the device code.

We're working with a few external folks such as those in this thread to provide some early guidance as to how it's done, and at some point we'll have some nice examples to point at (beyond say, our CPU and GPU implementations).

@RichDubielzig: the dataflow model of TF execution means that any operations that can be run at a given time will be run by the executor (assuming there are enough inter-op-parallelism threads to run them).  So yes, it is possible to have multiple "matmul" nodes running at once if both can be run at the same time.

On CPU, we use Eigen for a lot of the computation, and each op has a shared threadpool on which to execute -- so those threads will typically always be busy if there's lots of op work to do (though I admit, it probably won't be super efficient to context switch all the time, but I digress).

On GPU, we use GPU streams to enqueue asyncrhronous ops: execution of an OpKernel just queues the work to be done, and we use streams to enforce execution order at the actual device. 

@hawkinsp might have more to say here.

**Comment by CUinNYC** [comment]
@vrv I really appreciate the initial documentation. Thanks to that, I successfully created a "fake CPU" (really easy indeed). It required minor changes because of the changes into the TensorFlow code (r1.0). So having an external-module approach would be more portable. Let me know if you need a help or feedback comments on a preliminary documentation. 

In particular, I am particularly interested in how to allocate memory for accelerators. My goal is to reduce as much as possible memory copies.

If anybody has comments or examples I will really appreciate that.

**Comment by RichDubielzig** [comment]
Just following up on my question:  I am seeing results with this approach:  

1.  Define my operation as an AsyncOpKernel.  Kernel pseudocode:
```
ComputeAsync( context  input, DoneCallback done):
  Get input tensors A & B from context
  Also get a Knureon command queue handle Q from the DeviceContext object.
  Figure out the dimensions of the tensors
  Assemble a call f(A,B,dims) to my Knureon matrix multipllier algorithm
  Queue the tuple (f(), Q, done) in the launch queue.
  Return immediately
```

2. In launch queue, in a separate thread:
```
(as part of init routine, we have created our own dedicated ThreadPool, which includes a thread to run this wait/spawn loop)

  Wait for resources R to run f()
  schedule a task to run f() on R resources using Q and calling done().
```

3. In the task scheduled above:
```
  deploy f() to R and kick it off.
  wait on Q to finish
  release R back to the Knureon environment
  call the AsyncOpKernel done() callback
  exit
```

**Comment by RichDubielzig** [comment]
Asked a new question on StackOverflow about the ConstTensor, which doesn't map to the DMAHelper::base() function:

http://stackoverflow.com/questions/42707600/tensorflow-how-to-get-pointer-to-data-contents-of-consttensor

**Comment by RichDubielzig** [comment]
Posted another question on StackOverflow regarding my confusion with DeviceContext  and when it needs to be used.  I am revisiting the issue because it turns out we have a smaller limit than I thought on open queues with the system and I'm wondering if I should

1. redesign with the idea of mapping one DeviceContext per queue
2. keep with the current design of one DeviceContext per OpKernelContext and manage queues in a separate pool as they are needed.

http://stackoverflow.com/questions/42869340/tensorflow-why-does-a-gpu-device-only-have-one-device-context-and-does-it-rea

**Comment by RichDubielzig** [comment]
Another issue we have run into in attempting to debug XLA:  It doesn't seem like we are able to exercise any backend compiler when running in a debugger.  The issue is here:

https://github.com/tensorflow/tensorflow/issues/9194

**Comment by vrv** [comment]
@RichDubielzig I've been told that XLA-related development discussion has been moved to https://groups.google.com/forum/#!forum/xla-dev  -- you might get more help / feedback there now.

**Comment by RichDubielzig** [comment]
Asked a new question related to this on stackoverflow.  Will also post to the xla-dev forum

https://stackoverflow.com/questions/44271775/can-tensorflow-xla-be-supported-by-plug-ins

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by shivaniag** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
